src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -4200,6 +4203,38 @@ private IOperation ConvertToIDisposable(IOperation operand, ITypeSymbol iDisposa
         {
             StartVisitingStatement(operation);
 
+            // `lock (l) { }` on value of type `System.Threading.Lock` is lowered to `using (l.EnterLockScope()) { }`.
+            if (operation.LockedValue.Type?.IsLockType() == true)
+            {
+                var lockTypeInfo = operation.LockedValue.Type.TryFindLockTypeInfo();
+
+                var lockObject = VisitRequired(operation.LockedValue);
+
+                var enterLockScope = lockTypeInfo.EnterLockScopeMethod is { } targetMethod
+                    ? new InvocationOperation(
+                        targetMethod: targetMethod,
+                        constrainedToType: null,
+                        instance: lockObject,
+                        isVirtual: targetMethod.IsVirtual || targetMethod.IsAbstract || targetMethod.IsOverride,
+                        arguments: ImmutableArray<IArgumentOperation>.Empty,
+                        semanticModel: null,
+                        syntax: lockObject.Syntax,
+                        type: targetMethod.ReturnType,
+                        isImplicit: true)
+                    : MakeInvalidOperation(type: null, lockObject);
+
+                HandleUsingOperationParts(
+                    resources: enterLockScope,
+                    body: operation.Body,
+                    disposeMethod: lockTypeInfo.ScopeDisposeMethod,
```

Comment:
>lockTypeInfo.ScopeDisposeMethod [](http://example.com/codeflow?start=35&length=31)

It looks like if we pass `null` here, the ```HandleUsingOperationParts``` is going to attempt converting the resource to ```IDisposable```, which wouldn't be the right thing to do.